streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java

```diff
@@ -316,59 +312,103 @@ public void shouldLogAndMeterWhenSkippingExpiredWindowByGrace() {
-            )
-            .toStream()
-            .map((key, value) -> new KeyValue<>(key.toString(), value))
-            .to("output");
+        stream1.groupByKey(Grouped.with(Serdes.String(), Serdes.String()))
+               .windowedBy(TimeWindows.of(ofMillis(10)).advanceBy(ofMillis(10)).grace(ofMillis(90L)))
+               .aggregate(
+                   () -> "",
+                   MockAggregator.toStringInstance("+"),
+                   Materialized.<String, String, WindowStore<Bytes, byte[]>>as("topic1-Canonicalized").withValueSerde(Serdes.String()).withCachingDisabled().withLoggingDisabled()
+               )
+               .toStream()
+               .map((key, value) -> new KeyValue<>(key.toString(), value))
+               .to("output");
 
         LogCaptureAppender.setClassLoggerToDebug(KStreamWindowAggregate.class);
         final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
         try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props, 0L)) {
-            driver.pipeInput(recordFactory.create(topic, "k", "100", 100L));
-            driver.pipeInput(recordFactory.create(topic, "k", "0", 0L));
-            driver.pipeInput(recordFactory.create(topic, "k", "1", 1L));
-            driver.pipeInput(recordFactory.create(topic, "k", "2", 2L));
-            driver.pipeInput(recordFactory.create(topic, "k", "3", 3L));
-            driver.pipeInput(recordFactory.create(topic, "k", "4", 4L));
-            driver.pipeInput(recordFactory.create(topic, "k", "5", 5L));
+            driver.pipeInput(recordFactory.create(topic, "k", "100", 200L));
+            driver.pipeInput(recordFactory.create(topic, "k", "0", 100L));
+            driver.pipeInput(recordFactory.create(topic, "k", "1", 101L));
+            driver.pipeInput(recordFactory.create(topic, "k", "2", 102L));
+            driver.pipeInput(recordFactory.create(topic, "k", "3", 103L));
+            driver.pipeInput(recordFactory.create(topic, "k", "4", 104L));
+            driver.pipeInput(recordFactory.create(topic, "k", "5", 105L));
             driver.pipeInput(recordFactory.create(topic, "k", "6", 6L));
             LogCaptureAppender.unregister(appender);
 
-            final MetricName metricName = new MetricName(
-                "late-record-drop-total",
-                "stream-processor-node-metrics",
-                "The total number of occurrence of late-record-drop operations.",
-                mkMap(
-                    mkEntry("client-id", "topology-test-driver-virtual-thread"),
-                    mkEntry("task-id", "0_0"),
-                    mkEntry("processor-node-id", "KSTREAM-AGGREGATE-0000000001")
-                )
-            );
-            assertThat(driver.metrics().get(metricName).metricValue(), equalTo(7.0));
+            assertLatenessMetrics(driver, is(7.0), is(101.0), isRoughly(15));
```

Comment:
Ditto here: are we certain there is no flakiness introduced even with `isRoughly` approximation?

Reply:
see above